Made Metadata clients Fido compatible (#4358)
sunpy/sunpy: docs/guide/acquiring_data/hek.rst

************************

The Heliophysics Event Knowledgebase (HEK) is a repository of feature
and event information about the Sun.
Entries are generated both by automated algorithms and human observers.
SunPy accesses this information through the `~sunpy.net.hek` module, which was developed through support from the European Space Agency Summer of Code in Space (ESA-SOCIS) 2011.

1. A simple query
*****************

To search the HEK, you need a start time, an end time, and an event type.
Times are specified as strings or Python datetime objects.
Event types are specified as upper case, two letter strings, and are identical to the two letter abbreviations found at the HEK website, http://www.lmsal.com/hek/VOEvent_Spec.html.

    >>> from sunpy.net import attrs as a
    >>> from sunpy.net import Fido
    >>> tstart = '2011/08/09 07:23:56'
    >>> tend = '2011/08/09 12:40:29'
    >>> event_type = 'FL'
    >>> result = Fido.search(a.Time(tstart,tend), a.hek.EventType(event_type))  # doctest: +REMOTE_DATA

> The first line defines the search start and end times.

Reviewer: ```suggestion
``tstart`` and ``tend`` defines the start and end times of the query, and ``event_type`` specifies the event type which in this example we are searching for flares defined as ``FL``.
```

> The second line specifies the event type, in this 'FL' or flare.

Reviewer: ```suggestion

```

Line 6 goes out to the web, contacts the HEK, and queries it for the information you have requested.
Event data for ALL flares available in the HEK within the time range 2011/08/09 07:23:56 UT - 2011/08/09 12:40:20 UT will be returned, regardless of which feature recognition method used to detect the flare.

Let's break down the arguments of ``Fido.search``.
The first argument::

    a.Time(tstart,tend)

sets the start and end times for the query.

The second argument::

    a.hek.EventType(event_type)

sets the type of event to look for.
Since we have defined ``event_type = 'FL'``, this sets the query to look for flares.
We could have also set the flare event type using the syntax::

    a.hek.FL

There is more on the attributes of hek in section 3 of this guide.

2. The result
*************

So, how many flare detections did the query turn up?

>     >>> len(result[0])  # doctest: +REMOTE_DATA

Reviewer: maybe should mention why ``result[0]`` is indexed here

Using list comprehensions, it is easy to get a list of the feature recognition methods used to find each of the flares in the result object, for example:

    >>> [elem["frm_name"] for elem in result]  # doctest:+REMOTE_DATA
    [<HEKColumn name='frm_name' dtype='str32' length=19>
                              asainz
                              asainz
                              asainz
                              asainz
                              asainz
                              asainz
                              asainz
                   SSW Latest Events
                                SWPC
    Flare Detective - Trigger Module
    Flare Detective - Trigger Module
                                SWPC
                   SSW Latest Events
    Flare Detective - Trigger Module
    Flare Detective - Trigger Module
    Flare Detective - Trigger Module
    Flare Detective - Trigger Module
    Flare Detective - Trigger Module
    Flare Detective - Trigger Module]

It is likely each flare on the Sun was actually detected multiple times by many different methods.

3. More complex queries
***********************

The Fido allows you to make more complex queries.

> There are two key features you need to know in order to make use of the full power of the Fido.

Reviewer 2: Fixed.

Firstly, the attribute module - attrs.hek - describes ALL the parameters stored by the HEK as listed in http://www.lmsal.com/hek/VOEvent_Spec.html, and the HEK client makes these parameters searchable.

> To explain this, let's have a closer look at attrs.hek.

Reviewer 2: Fixed.

SunPy's HEK module is in sunpy.net.  It can be imported into your
session as follows:

>     >>> from sunpy.net import attrs as a, hek

Reviewer 3: Does this example use `hek` anywhere now?
```suggestion
    >>> from sunpy.net import attrs as a
```

>     >>> client = hek.HEKClient()  # doctest: +REMOTE_DATA

Reviewer 2: Do we want to update this with fido instead now?

Reviewer 4: Agreed

    >>> tstart = '2011/08/09 07:23:56'
    >>> tend = '2011/08/09 12:40:29'
    >>> event_type = 'FL'

>     >>> result = client.search(a.Time(tstart,tend),a.hek.EventType(event_type))  # doctest: +REMOTE_DATA

Reviewer 3: ```suggestion
    >>> result = client.search(a.Time(tstart,tend), a.hek.EventType(event_type))  # doctest: +REMOTE_DATA
```

Since we are using an ```and`` logical operator we need a result from the ``(a.hek.FL.PeakFlux > 1000.0)`` filter.
Flares that have `None` for a peak flux cannot provide this, and so are excluded.
The `None` type in this context effectively means "Don't know"; in such cases the client returns only those results from the HEK that definitely satisfy the criteria passed to it.

4. Getting data for your event
******************************

The 'hek2vso' module allows you to take an HEK event and acquire VSO records specific to that event and was developed with support from the 2013 Google Summer of Code.

    >>> from sunpy.net import hek2vso
    >>> h2v = hek2vso.H2VClient()  # doctest: +REMOTE_DATA

There are several ways to use this capability.
For example, you can pass in a list of HEK results and get out the corresponding VSO records.
Here are the VSO records returned via the tenth result from the HEK query in Section 2 above:

    >>> result = Fido.search(a.Time(tstart,tend), a.hek.EventType(event_type))  # doctest: +REMOTE_DATA
    >>> vso_records = h2v.translate_and_query(result[0][10])  # doctest: +REMOTE_DATA
    >>> len(vso_records[0])  # doctest: +REMOTE_DATA
    31

> Result 10 is an HEK entry generated by the "Flare Detective" automated flare detection algorithm running on the AIA 193 angstrom waveband.

Reviewer: what is `Result 10`?

Reviewer: ```suggestion
``result[0][10] is an HEK entry generated by the "Flare Detective" automated flare detection algorithm running on the AIA 193 angstrom waveband.
```

> The VSO records are for full disk AIA 193 images between the start and end times of this event.

Reviewer: ```suggestion
The VSO records are for full disk AIA 193 angstrom images between the start and end times of this event.
```

Latest Events tool.  I can retrieve those entries only from the HEK
with the following command:

>     >>> result = client.search( a.Time(tstart,tend), a.hek.EventType(event_type), a.hek.FRM.Name == 'SSW Latest Events')  # doctest: +REMOTE_DATA

Reviewer 3: ```suggestion
    >>> result = client.search(a.Time(tstart,tend), a.hek.EventType(event_type), a.hek.FRM.Name == 'SSW Latest Events')  # doctest: +REMOTE_DATA
```

    >>> len(result)  # doctest: +REMOTE_DATA
    2

We can also retrieve all the entries in the time range which were not
made by SSW Latest Events with the following command:

>     >>> result = client.search( a.Time(tstart,tend), a.hek.EventType(event_type),a.hek.FRM.Name != 'SSW Latest Events')  # doctest: +REMOTE_DATA

Reviewer 3: ```suggestion
    >>> result = client.search(a.Time(tstart,tend), a.hek.EventType(event_type), a.hek.FRM.Name != 'SSW Latest Events')  # doctest: +REMOTE_DATA
```

caution is advisable.  Let's say I want all the flares west of 50
arcseconds OR have a peak flux over 1000.0:

>     >>> result = client.search(a.Time(tstart,tend), a.hek.EventType(event_type), (a.hek.Event.Coord1 > 50) or (a.hek.FL.PeakFlux > 1000.0) )  # doctest: +REMOTE_DATA

Reviewer 3: ```suggestion
    >>> result = client.search(a.Time(tstart,tend), a.hek.EventType(event_type), (a.hek.Event.Coord1 > 50) or (a.hek.FL.PeakFlux > 1000.0))  # doctest: +REMOTE_DATA
```

Let's say we want all the flares west of 50 arcseconds AND have a peak
flux over 1000.0:

>     >>> result = client.search(a.Time(tstart,tend), a.hek.EventType(event_type), (a.hek.Event.Coord1 > 50) and (a.hek.FL.PeakFlux > 1000.0) )  # doctest: +REMOTE_DATA

Reviewer 3: ```suggestion
    >>> result = client.search(a.Time(tstart,tend), a.hek.EventType(event_type), (a.hek.Event.Coord1 > 50) and (a.hek.FL.PeakFlux > 1000.0))  # doctest: +REMOTE_DATA
```

records.  Here are the VSO records returned via the tenth result from
the HEK query in Section 2 above:

>     >>> result = client.search(a.Time(tstart,tend),a.hek.EventType(event_type))  # doctest: +REMOTE_DATA

Reviewer 3: ```suggestion
    >>> result = client.search(a.Time(tstart,tend), a.hek.EventType(event_type))  # doctest: +REMOTE_DATA
```
